pkg/ipam-controller/allocator/allocator.go

```diff
@@ -115,76 +93,90 @@ func (pa *poolAllocator) Allocate(ctx context.Context, node string) (nodeAllocat
 		ip.IsBroadcast(endIP, pa.cfg.Subnet) {
 		// out of range
 		log.Info("can't allocate: pool has no free ranges")
-		return nodeAllocationInfo{}, ErrNoFreeRanges
+		return &AllocatedRange{}, ErrNoFreeRanges
 	}
 
 	log.Info("range allocated",
 		"start", startIP, "end", endIP)
-	pa.allocations[node] = allocatedRange{
+	a := AllocatedRange{
 		StartIP: startIP,
 		EndIP:   endIP,
 	}
-	return pa.getNodeAllocationInfo(node, pa.allocations[node]), nil
+	pa.allocations[node] = a
+	return &a, nil
+}
+
+// AllocateFromPool allocates ranges for node
+func (pa *PoolAllocator) AllocateFromPool(ctx context.Context, nodeName string) (*AllocatedRange, error) {
+	allocation, err := pa.allocate(ctx, nodeName)
+	if err != nil {
+		pa.Deallocate(ctx, nodeName)
+		return nil, err
+	}
+	return allocation, nil
 }
 
 // Deallocate remove info about allocation for the node from the poolAllocator
-func (pa *poolAllocator) Deallocate(ctx context.Context, node string) {
+func (pa *PoolAllocator) Deallocate(ctx context.Context, node string) {
 	log := pa.getLog(ctx, pa.cfg)
 	log.Info("deallocate range for node", "node", node)
 	delete(pa.allocations, node)
 }
 
-// Load loads range to the pool allocator with validation for conflicts
-func (pa *poolAllocator) Load(ctx context.Context, allocData nodeAllocationInfo) error {
-	log := pa.getLog(ctx, pa.cfg).WithValues("node", allocData.Node)
-	if err := pa.checkAllocation(allocData); err != nil {
+// load loads range to the pool allocator with validation for conflicts
+func (pa *PoolAllocator) load(ctx context.Context, nodeName string, allocRange AllocatedRange) error {
+	log := pa.getLog(ctx, pa.cfg).WithValues("node", nodeName)
+	if err := pa.checkAllocation(allocRange); err != nil {
 		log.Info("range check failed", "reason", err.Error())
 		return err
 	}
+	a, ok := pa.allocations[nodeName]
+	if ok && allocRange.StartIP.Equal(a.StartIP) {
+		// Node Allocation already exists and is the same
+		return nil
+	}
 	allocations := pa.getAllocationsAsSlice()
 	for _, a := range allocations {
 		// range size is always the same, then an overlap means the blocks are necessarily equal.
 		// it's enough to just compare StartIP which can technically act as an absolute "block index" in the subnet
-		if allocData.allocatedRange.StartIP.Equal(a.StartIP) {
+		if allocRange.StartIP.Equal(a.StartIP) {
```

Comment:
This is not directly related to your changes, but I think we should address this now. Check for overlaps is implemented in an inefficient manner. Before your changes, we were calling the load logic only on controller start, but now it is invoked on each iteration for each existing allocation, and we should optimize it. I don't like the fact that we need to do a linear search for each allocation that we are trying to load. We could probably create an extra internal index (map) that will contain the startIPs, and we may use it for overlap checks. We will need to update this index in each place where we modify `pa.allocations` map. WDYT?

Reply:
Done